### ISSUE TITLE
Having lots of failed payments: WIRE_FEE_INSUFFICIENT

### ISSUE THREAD

**Issue by billygarrison** [ElementsProject/lightning#2362]
### Issue and Steps to Reproduce
I am writing a script to do circular payments. It doesn't yet use accurate delay/fees, so it just adds 200 blocks to the delay for each hop. It adds 1100 msatoshis to each hop for routing fees, as I'm testing with small amounts (e.g. 1000 msats), and most nodes have a base fee of 500 to 1000 msats. So in theory this should work most of the time.

I've had it work once. The rest of the times I get errors. Sometimes its `WIRE_TEMPORARY_CHANNEL_FAILURE` but usually it's `WIRE_FEE_INSUFFICIENT` (e.g. below). I've checked the node / channel on 1ml.com, and so far they've always been one of LNBIG's nodes. When looking at the channel info on that site, they seem to have very low fees, so I'm not sure why the fees I add aren't working. I've also tried upping the fee to 10,000 msats per hop just in case, and it still doesn't work.
```
    {
      "type": "INFO",
      "time": "224886.434847165",
      "source": "lightningd(6634):",
      "log": "0237fefbe8626bf888de0cad8c73630e32746a22a2c4faa91c1d9877a3826e1174 chan #10: htlc 26 failed from 2th node with code 0x100c (WIRE_FEE_INSUFFICIENT)"
    },
```

I have heard that LNBIG is not a good node for routing, so perhaps this has to do with that? Should I write my script to avoid LNBIG's nodes, or might I be doing something wrong?

Here's an example of the route I construct with 200 block delay and 10k msats fee per hop, maybe there's an error there:
```
[
  {
    "msatoshi": 1020013,
    "delay": 800,
    "id": "0237fefbe8626bf888de0cad8c73630e32746a22a2c4faa91c1d9877a3826e1174",
    "channel": "561410:108:1"
  },
  {
    "msatoshi": 1010013,
    "delay": 600,
    "id": "03ee180e8ee07f1f9c9987d98b5d5decf6bad7d058bdd8be3ad97c8e0dd2cdc7ba",
    "channel": "554084:766:0"
  },
  {
    "msatoshi": 1010000,
    "delay": 400,
    "id": "02c12b5459cf107ee0440cae41902f1189db50fa003a077f3a6fbe6b5760218695",
    "channel": "557647:1640:0"
  },
  {
    "msatoshi": 1000000,
    "delay": 200,
    "id": "02843833374b6c42daa4b780df2066a9985cb8cf97bc4566dd63cf693d32cbf7b5",
    "channel": "561084:1157:1"
  }
]
```

### `getinfo` output
```
{
  "id": "02843833374b6c42daa4b780df2066a9985cb8cf97bc4566dd63cf693d32cbf7b5",
  "alias": "BillyGarrison",
  "color": "009900",
  "num_peers": 12,
  "num_pending_channels": 2,
  "num_active_channels": 10,
  "num_inactive_channels": 0,
  "address": [
    {
      "type": "ipv4",
      "address": "47.55.123.202",
      "port": 9735
    }
  ],
  "binding": [
  ],
  "version": "v0.6.3",
  "blockheight": 563541,
  "network": "bitcoin",
  "msatoshi_fees_collected": 3825
}
```

### EDIT
A couple things I should add. I retried today and it failed at a non-LNBIG node (although it is a big node). This node and LNBIG seem to have all of their channel fees 1000 msat or lower, with very low fee rate. Also it may be worth mentioning that this script was only successful the very first time it ran, and the 30 times since then it has failed. 

Here's how I create the route:
- Calculate amount each channel is unbalanced by
- Say I have 2 channels, one is +100,000 sats on the local side and the other is -200,000 on the remote side. The script chooses to pay 100,000 through the +100,000 channel (I call that node payThroughNode), and receives through the -200,000 channel (receiveThroughNode)
- I use `node.getroute(receiveThroughNodeID, rebalanceAmount*1000, 5, 20, payThroughNodeID, "", 0)`, the extra parameters are `riskfactor=5`, `cltv=20`, `fuzzpercent=0`.
- Then a function adds a hop on either end (from me to payThroughNode and from receiveThroughNode to me).
- I give each hop a delay of an extra 200. And I add 1100 msat to fees (and have tried adding 10k msat) for each hop.

I get `payment_hash` by creating a bolt11 invoice for `rebalanceAmount*1000` (`rebalanceAmount` is in satoshis), and using bolt11["payment_hash"]. I then use `sendpay` to send the payment along the route.

I also tried rolling the script back to a commit I made immediately after the successful payment, and it still failed with the same `WIRE_FEE_INSUFFICENT` error.

**Comment by billygarrison** [comment]
I have a couple more questions regarding this script.

If I want to know what delay and fees to use for a hop, is it okay to use `listchannels` and get the `delay`,  `base_fee_millisatoshi`m and `fee_per_millisatoshi` values?

And how do I know which to use, since `listchannels` gives two results for each channel? Is it just based on the source / destination? So if I want to pay from my node through `payThroughNode`, in the following example which base fee, fee rate, and delay would I need to use for the first hop (me to payThroughNode)?

Example:
```
    {
      "source": "myNodeID",
      "destination": "payThroughNodeID",
      "short_channel_id": "535391:1083:0",
      "public": true,
      "satoshis": 400000,
      "message_flags": 0,
      "channel_flags": 2,
      "flags": 2,
      "active": false,
      "last_update": 1549908418,
      "base_fee_millisatoshi": 1000,
      "fee_per_millionth": 1,
      "delay": 144
    },
    {
      "source": "payThroughNodeID",
      "destination": "myNodeID",
      "short_channel_id": "535391:1083:0",
      "public": true,
      "satoshis": 400000,
      "message_flags": 0,
      "channel_flags": 3,
      "flags": 3,
      "active": false,
      "last_update": 1549978194,
      "base_fee_millisatoshi": 0,
      "fee_per_millionth": 1,
      "delay": 144
    },
```

**Comment by rustyrussell** [comment]
Hey, great to see you playing with manual routes!  I thought I was the only one!

The reason you get two results is that each direction of the channel is controlled independently.  (Note how source and dest are reversed).  You need to use the right direction: in current master you also get a direction flag explicitly.

BTW, you want to use the formula from BOLT 7: https://github.com/lightningnetwork/lightning-rfc/blob/master/07-routing-gossip.md#htlc-fees

And you need to calculate it *backwards* since every channel needs to carry the fees sufficient for the next channel...

Now, as to your exact case: you're giving a 13 msat fee to 03ee180e8ee07f1f9c9987d98b5d5decf6bad7d058bdd8be3ad97c8e0dd2cdc7ba for channel 554084:766:0.  But that direction has "base_fee_millisatoshi": 150, "fee_per_millionth": 2.

**Comment by billygarrison** [comment]
Thanks for the feedback @rustyrussell! I can't believe I didn't see the 13 msat fee... I was so focused on the first/last hop that I manually constructed, that I didn't think to scrutinize over the ones from `getroute`. I'll take some more time to look at the results from each intermediate step.

To determine the correct direction, am I right to assume that a hop from A to B means A is source and B is destination?

And is the `delay` in `listchannels` output the correct one to use in the route (I understand I must add it to the previous delays).

Last question - thanks for sharing the formula from BOLT 7, is it absolutely necessary to follow the formula, or is it okay to over-estimate? (For testing purposes)

I did know about the formula, but was trying to overpay in fees/delay to make sure it worked. Because I was getting an error that the payment would get to the recipient with too little time to spare, so I upped the delays and it fixed that.

**Comment by billygarrison** [comment]
I've re-written the script to calculate delays/fees based on `listchannels` info for the peer, and now I get the following failure almost every time (not always the 0th node)

```
    {
      "type": "INFO",
      "time": "134181.155293089",
      "source": "lightningd(5589):",
      "log": "032a57d7e327f5be21d4fdc12b171b49a75a19b34af428699a7dcc06a9889ffccb chan #30: htlc 2 failed from 0th node with code 0x1007 (WIRE_TEMPORARY_CHANNEL_FAILURE)"
    },
```

I tried searching this repo + the BOLT repo to find info about that failure message but couldn't find much.

It seems that lower value payments seem to have a higher likelihood of succeeding (9000 sats was the highest to succeed, although it doesn't succeed every time). 1 sat seems to work (almost) every time. The script only takes ~1 second to run so it seems unlikely that it's due to some sort of disconnection between the time it uses `getroute` and the time it uses `sendpay`.

When a payment succeed, there is a delay of ~3 seconds for it to go from `pending` to `complete` in `listpayments` output. When it fails, it changes to `failed` nearly instantly.

I wonder if it could have anything to do with the fact that it just uses `getroute` once and then tries `sendpay` once?

The circular route looks like:
```
[
  {
    "msatoshi": 80001739,
    "delay": 441,
    "id": "03555a11db2e0c3a97571ba144076adac8513d68a2cdf19580e2dcc54b25b78a3a",
    "channel": "563895:1346:0"
  },
  {
    "msatoshi": 80000659,
    "delay": 297,
    "id": "02ad6fb8d693dc1e4569bcedefadf5f72a931ae027dc0f0c544b34c1c6f3b9a02b",
    "channel": "551142:1544:0"
  },
  {
    "msatoshi": 80000579,
    "delay": 153,
    "id": "02c12b5459cf107ee0440cae41902f1189db50fa003a077f3a6fbe6b5760218695",
    "channel": "555782:926:1"
  },
  {
    "msatoshi": 80000000,
    "delay": 9,
    "id": "02843833374b6c42daa4b780df2066a9985cb8cf97bc4566dd63cf693d32cbf7b5",
    "channel": "561084:1157:1"
  }
]
```

**Comment by billygarrison** [comment]
Another update: I looped through all of my channels to try doing a circular payment through each. Two of them worked for 50k sats and the others failed. I ran the script again and the same 2 nodes routed 50k sats again.

The two that worked:
https://1ml.com/channel/618906299264598017
https://1ml.com/channel/619341705884663809

Two examples that didn't:
https://1ml.com/channel/619605588663140353
https://1ml.com/channel/615635252158988288
